web_tree_many2one_clickable/static/src/js/web_tree_many2one_clickable.js

```diff
@@ -1,104 +1,112 @@
 
-openerp.web_tree_many2one_clickable = function(instance, local)
-{
-    instance.web.list.Column.include({
-        /*
-        Load config parameter at init and store it in an accessible variable.
-        */
-        init: function(id, tag, attrs) {
-            this._super(id, tag, attrs);
-            if (this.widget == 'many2one_clickable') {
-                this.use_many2one_clickable = true;
-            } else if (this.type == 'many2one') {
-                this.get_options();
-            }
-        },
+odoo.define('web_tree_many2one_clickable.many2one_clickable', function(require) {
+"use strict";
 
-        get_options: function() {
-            if (_.isUndefined(this.ir_option_clickable_loaded)) {
-                var self = this; // Needed for binding the instance
-                this.ir_option_clickable_loaded = $.Deferred();
-                this.use_many2one_clickable = false;
-                (new instance.web.Model("ir.config_parameter"))
-                    .query(["value"])
-                    .filter([['key', '=', 'web_tree_many2one_clickable.default']])
-                    .first()
-                    .then(function(param) {
-                        if (param) {
-                            self.use_many2one_clickable = (param.value.toLowerCase() == 'true');
-                        }
-                        self.ir_option_clickable_loaded.resolve();
-                    });
-                return this.ir_option_clickable_loaded;
-            }
-            return $.when();
-        },
+var _ = require('_');
+var $ = require('$');
+
+var core = require('web.core');
+var ListView = require('web.ListView');
+var Model = require('web.DataModel');
 
-        _format: function (row_data, options)
-        {
-            if (this.use_many2one_clickable) {
-                return _.str.sprintf('<a class="oe_form_uri" data-many2one-clickable-model="%s" data-many2one-clickable-id="%s">%s</a>',
-                    this.relation,
-                    row_data[this.id].value[0],
-                    _.escape(row_data[this.id].value[1] || options.value_if_empty));
+var _t = core._t;
+var QWeb = core.qweb;
+var list_widget_registry = core.list_widget_registry;
+
+var use_many2one_clickable_default = false;
+var ir_option_clickable_loaded;
+
+function get_options() {
+    if (_.isUndefined(ir_option_clickable_loaded)) {
+        ir_option_clickable_loaded = $.Deferred();
+        (new Model("ir.config_parameter"))
+            .query(["value"])
+            .filter([['key', '=', 'web_tree_many2one_clickable.default']])
+            .first()
+            .then(function(param) {
+                if (param) {
+                    use_many2one_clickable_default = (
+                        param.value.toLowerCase() == 'true');
+                }
+                ir_option_clickable_loaded.resolve();
+            });
+        return ir_option_clickable_loaded;
+    }
+    return $.when();
+}
```

Comment:
If I'm not missing something, this is quite overengineered. Let me explain. You currently do this:

1. [Create a `false` default value](https://github.com/OCA/web/pull/456/files#diff-f06f62892caad3befc7bded8af2e3c50R20). This is weird, because instead of having defaults, you should subscribe to the promise.
2. [Create an empty variable](https://github.com/OCA/web/pull/456/files#diff-f06f62892caad3befc7bded8af2e3c50R21) that is expected to [become a promise later](https://github.com/OCA/web/pull/456/files#diff-f06f62892caad3befc7bded8af2e3c50R25). This is weird because you should define it in the 1st step and simply reject or resolve it later.
3. [Query `ir.config_parameter` manually](https://github.com/OCA/web/pull/456/files#diff-f06f62892caad3befc7bded8af2e3c50R27) and check for results, weird because you could simply call its `get_param` method, that already includes the possibility to have a default value.
4. [The returned promise](https://github.com/OCA/web/pull/456/files#diff-f06f62892caad3befc7bded8af2e3c50R39) only tells you the parameter got loaded, which is weird because normally you are expected to get a result from the promise.
5. [There is a check to know what promise to return](https://github.com/OCA/web/pull/456/files#diff-f06f62892caad3befc7bded8af2e3c50R24), but why do you return different promises instead of just resolving or rejecting the one you should already have declared?
6. Later in the stack, [you check if there was a positive default](https://github.com/OCA/web/pull/456/files#diff-f06f62892caad3befc7bded8af2e3c50R48) and [then apply it](https://github.com/OCA/web/pull/456/files#diff-f06f62892caad3befc7bded8af2e3c50R49), and [try to fetch the default if it is negative](https://github.com/OCA/web/pull/456/files#diff-f06f62892caad3befc7bded8af2e3c50R51). This is wierd too, because either you could reach that line before the promise is resolved and you are not applying the value when it does, or you could get there after resolution, but the returned parameter was `false` and then you'd be calling `get_options` all the time. Yes, [it checks for the existence of the `ir_option_clickable_loaded` variable](https://github.com/OCA/web/pull/456/files#diff-f06f62892caad3befc7bded8af2e3c50R24) to avoid asking the server again, but that is again not required if you do my solution because when a promise is already resolved and you attach a handler later to it, it gets immediatly executed automatically.
7. You are not exposing this method nor the promise, so this is not really inheritable from another addon.

So yes, this is very weird as it is right now; when doing async programming there are these things to keep in mind.

Mi suggestion avoids the `use_many2one_clickable_default` + `ir_option_clickable_loaded` combination; it simply sets a promise called `clickable_default` that gets evaluated only once when the module is loaded and that is correctly resolved with a value. Besides, it is much shorter.

If you still want to make it inheritable, which is good, then simply expose the promise and other addons will be able to subscribe to it.